### PR TITLE
BuildKit (May 18, 2018) + massive dedupe (with local session mgr)

### PR DESCRIPTION
**Pull request by AkihiroSuda**
Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

* Vendors the latest BuildKit (with "llbsolver-next")
* Introduce loopback session manager using https://github.com/moby/buildkit/blob/master/session/testutil/testutil.go
* Massive dedupe with BuildKit

- - -

Vendor

* [x] https://github.com/moby/buildkit/pull/393

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
@jessfraz @tonistiigi

File: client/session.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		var err error
+		c.sessionManager, err = session.NewManager()
+		if err != nil {
+			return nil, err
+		}
+	}
+	return c.sessionManager, nil
+}
+
+func (c *Client) Session(ctx context.Context) (*session.Session, session.Dialer, error) {
```
Comment (AkihiroSuda) [inline client/session.go]: Please review this file.
Other changes are mostly just vendoring updates.

**Comment by AkihiroSuda** [comment]
CI failing
```
=== RUN   TestPullIsInListOutput
--- FAIL: TestPullIsInListOutput (0.68s)
	pull_test.go:29: expected busybox:latest in ls output, got: calculating size of image docker.io/library/testbuilddockerfilenotincontext:latest failed: content digest sha256:145ea632b218505d8dfa87871ec80566cc74eda77adf7ea6f252a3f01f472956: not found
=== RUN   TestRemoveBuiltImage
--- FAIL: TestRemoveBuiltImage (0.72s)
	remove_test.go:19: expected testremoveimage in ls output, got: calculating size of image docker.io/library/testbuilddockerfilenotincontext:latest failed: content digest sha256:145ea632b218505d8dfa87871ec80566cc74eda77adf7ea6f252a3f01f472956: not found
=== RUN   TestRemovePulledImage
--- FAIL: TestRemovePulledImage (1.09s)
	remove_test.go:39: expected debian:buster in ls output, got: calculating size of image docker.io/library/testbuilddockerfilenotincontext:latest failed: content digest sha256:145ea632b218505d8dfa87871ec80566cc74eda77adf7ea6f252a3f01f472956: not found
=== RUN   TestSaveImage
--- PASS: TestSaveImage (1.32s)
=== RUN   TestTagImage
--- FAIL: TestTagImage (1.32s)
	tag_test.go:19: expected ls output to have tagthing:latest and jess/tagtest:latest but got: calculating size of image docker.io/library/testbuilddockerfilenotincontext:latest failed: content digest sha256:145ea632b218505d8dfa87871ec80566cc74eda77adf7ea6f252a3f01f472956: not found
FAIL
FAIL	github.com/genuinetools/img	23.531s
```

One of the failure reason is that `buildkit/source/containerimage/pull.go` lacks `images.Create`

https://github.com/genuinetools/img/blob/d8dcd33c70dd7b931508be77c582cce4aa8a3fba/source/containerimage/pull.go#L206-L222

**Comment by jessfraz** [comment]
This is amazing thanks so much, design LGTM

**Comment by tonistiigi** [comment]
@AkihiroSuda This is only for the "pull command" of img, right? Not for implicit pulls happening during build. I don't think it is a good idea to create images for implicit pulls - it is a source or cleanup issues and wouldn't work with containerd gc. Otoh it would be nice if the cache keys of these implicit pulls are stored so offline builds using them work. Would definitely like a solution for that in `buildkit/source/containerimage/pull.go`.

**Comment by AkihiroSuda** [comment]
> This is only for the "pull command" of img, right?

Yes.
Probably rather than putting `images.Create()` to `buildkit/source/containerimage/pull.go`, we should better add OCI descriptor to  the metadata of `ImmutableRef`, so that `img/client/pull.go` can call `images.Create()` with the metadata?

**Comment by tonistiigi** [comment]
@AkihiroSuda We can just make a shared pull function, similar to https://godoc.org/github.com/containerd/containerd#Client.Pull but depend on the individual store interfaces instead of a containerd client and with support for attaching progress. This would be called then by both image source and pull command.

File: Gopkg.lock [hunk 2]
```diff
@@ -345,17 +344,16 @@
   revision = "2bca23e0e452137f789efbc8610126fd8b94f73b"
 
 [[projects]]
-  branch = "master"
+  branch = "split-pull"
```
Comment (AkihiroSuda) [inline Gopkg.lock]: https://github.com/moby/buildkit/pull/393

**Comment by AkihiroSuda** [comment]
`img pull` seems working now but CI is still failing..
Looks some GC issue.

**Comment by AkihiroSuda** [comment]
Vendor PR is merged and CI is green

**Comment by jessfraz** [comment]
Awesome!!!

**Comment by jessfraz** [comment]
On a plane but will try and get enough internet to make a release thank you so so much for this

**Pull request merged**